File: epi_judge_java/epi/PivotList.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;
import epi.test_framework.TimedExecutor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PivotList {

    public static ListNode<Integer> listPivoting(ListNode<Integer> head, int x) {
        var leftDummy = new ListNode<>(-1, null);
        var rightDummy = new ListNode<>(-1, null);
        var midDummy = new ListNode<>(-1, null);

        var leftRunner = leftDummy;
        var rightRunner = rightDummy;
        var midRunner = midDummy;

        var runner = head;
        while (runner != null) {
            if (runner.data < x) {
                leftRunner.next = runner;
                leftRunner = leftRunner.next;
            } else if (runner.data == x) {
                midRunner.next = runner;
                midRunner = midRunner.next;
            } else {
                rightRunner.next = runner;
                rightRunner = rightRunner.next;
            }

            runner = runner.next;
        }

        if (midRunner != midDummy) {
            leftRunner.next = midDummy.next;
            midRunner.next = rightDummy.next;
            rightRunner.next = null;
        } else {
            leftRunner.next = rightDummy.next;
            rightRunner.next = null;
        }

        return leftDummy.next;
    }

    public static List<Integer> linkedToList(ListNode<Integer> l) {
        List<Integer> v = new ArrayList<>();
        while (l != null) {
            v.add(l.data);
            l = l.next;
        }
        return v;
    }

    @EpiTest(testDataFile = "pivot_list.tsv")
    public static void listPivotingWrapper(TimedExecutor executor,
                                           ListNode<Integer> l, int x)
            throws Exception {
        List<Integer> original = linkedToList(l);

        final ListNode<Integer> finalL = l;
        l = executor.run(() -> listPivoting(finalL, x));

        List<Integer> pivoted = linkedToList(l);

        int mode = -1;
        for (Integer i : pivoted) {
            switch (mode) {
                case -1:
                    if (i == x) {
                        mode = 0;
                    } else if (i > x) {
                        mode = 1;
                    }
                    break;
                case 0:
                    if (i < x) {
                        throw new TestFailure("List is not pivoted");
                    } else if (i > x) {
                        mode = 1;
                    }
                    break;
                case 1:
                    if (i <= x) {
                        throw new TestFailure("List is not pivoted");
                    }
            }
        }

        Collections.sort(original);
        Collections.sort(pivoted);
        if (!original.equals(pivoted))
            throw new TestFailure("Result list contains different values");
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "PivotList.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/pivot_list.cc
#include <algorithm>
#include <iterator>
#include <memory>
#include <vector>

#include "list_node.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"

using std::shared_ptr;

shared_ptr<ListNode<int>> ListPivoting(const shared_ptr<ListNode<int>> &list, int x) {
    shared_ptr<ListNode<int>> lessDummy(new ListNode<int>);
    shared_ptr<ListNode<int>> eqDummy(new ListNode<int>);
    shared_ptr<ListNode<int>> greaterDummy(new ListNode<int>);

    auto lessRunner = lessDummy;
    auto eqRunner = eqDummy;
    auto greaterRunner = greaterDummy;

    auto runner = list;
    while (runner != nullptr) {
        if (runner->data < x) {
            lessRunner->next = runner;
            lessRunner = lessRunner->next;
        } else if (runner->data == x) {
            eqRunner->next = runner;
            eqRunner = eqRunner->next;
        } else {
            greaterRunner->next = runner;
            greaterRunner = greaterRunner->next;
        }

        runner = runner->next;
    }

    greaterRunner->next = nullptr;
    eqRunner->next = greaterDummy->next;
    lessRunner->next = eqDummy->next;

    return lessDummy->next;
}

std::vector<int> ListToVector(const shared_ptr<ListNode<int>> &l) {
    std::vector<int> v;
    ListNode<int> *it = l.get();
    while (it) {
        v.push_back(it->data);
        it = it->next.get();
    }
    return v;
}

void ListPivotingWrapper(TimedExecutor &executor,
                         const shared_ptr<ListNode<int>> &l, int x) {
    std::vector<int> original = ListToVector(l);

    std::shared_ptr<ListNode<int>> pivoted_list =
            executor.Run([&] { return ListPivoting(l, x); });

    std::vector<int> pivoted = ListToVector(pivoted_list);
    enum class Mode {
        kLess, kEq, kGreater
    } mode = Mode::kLess;
    for (auto &i: pivoted) {
        switch (mode) {
            case Mode::kLess:
                if (i == x) {
                    mode = Mode::kEq;
                } else if (i > x) {
                    mode = Mode::kGreater;
                }
                break;
            case Mode::kEq:
                if (i < x) {
                    throw TestFailure("List is not pivoted");
                } else if (i > x) {
                    mode = Mode::kGreater;
                }
                break;
            case Mode::kGreater:
                if (i <= x) {
                    throw TestFailure("List is not pivoted");
                }
        }
    }
    std::sort(begin(original), end(original));
    std::sort(begin(pivoted), end(pivoted));
    if (original != pivoted) {
        throw TestFailure("Result list contains different values");
    }
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"executor", "l", "x"};
    return GenericTestMain(args, "pivot_list.cc", "pivot_list.tsv",
                           &ListPivotingWrapper, DefaultComparator{},
                           param_names);
}
